### DISCUSSION TITLE
Code generation with IPOPT solver

### DISCUSSION THREAD

**Discussion by Kamal-a-1** [casadi/casadi#3361] [Q&A]
Hello,

I wanted to ask, is there any way to generate C code from a MATLAB code involving "ipopt" solver in order to run it on a hardware in real-time?

I tried to generate dependencies to a .dll file, then use `ocp.solver = casadi.nlpsol('solver', 'ipopt', './sol.dll')` and then convert it using MATLAB coder, but MATLAB coder doesn't support `casadi.nlpsol()`.

Is there any workaround or is changing the solver the only option?

Thanks very much,
Kamal

**Comment by brunoml5** [comment]
Hi! I don't believe casadi is able to autogenerate code for `ipopt`. I think `sqpmethod` with qp solver such as `osqp` or `qrqp` works.

But if you interested in solving optimization in real time, take a look at [acados](https://github.com/acados/acados). They have special algorithms for this.

**Comment by espee** [comment]
It does look like casadi now supports ipopt method. Here is an example. https://web.casadi.org/blog/mpc-simulink2/
However, as I modify the codes for my case, I see error during evaluation of the casadi function. (Line 171 of casadi_fun.c). Without any further description about the failure to evaluate the result, I am stuck without any luck.

**Reply by farides86** [reply]
Hi @espee , its been almost a year, any luck for you to implement this? I'm still stuck with it, like you said, I have to modify the casadi_fun.c too. I'm using CasADi and IPOPT for my controller & now its time to do in real time where Speedgoat will be the target machine but there are couple of errors such as,
1) Error C1083: Cannot open include file: 'casadi/casadi_c.h': No such file or directory - Solved by added the path
1) Error C2065: 'mxIsChar_is_not_supported_in_Simulink_Coder', this is where I need to modify the casadi_fun.c file
2) Error LNK2019: Unresolved external symbol _casadi_c_push_file referenced in function _mdlInitializedSizes and others
3) Warning LNK4272: library machine type x64 conflicts with target machine type x86 (At this time, the Speedgoat was in 32 bits) - Stuck here 
4) Then I upgraded the Speedgoat system but there is incompatibility issue of Casadi library with Speedgoat architecture since it is a QNX based x64 bits system - then totally no idea what else to do.

No luck until now. Hopefully you manage to run it. Thank you
